compiler/noirc_frontend/src/parser/errors.rs

```diff
@@ -32,9 +32,11 @@ pub enum ParserErrorReason {
     #[error("Where clauses are allowed only on functions with generic parameters")]
     WhereClauseOnNonGenericFunction,
     #[error(
-        "Multiple primary attributes found. Only one primary attribute is allowed per function."
+        "Multiple primary attributes found. Only one primary attribute is allowed per function"
     )]
     MultiplePrimaryAttributesFound,
+    #[error("A primary attribute cannot be placed on a struct")]
+    NoPrimaryAttributesAllowedOnStruct,
```

Comment:
Why wouldn't primary attributes be allowed on a struct? This seems to signal a naming issue to me. Perhaps they should be function attributes if they're only allowed on functions for example.

Reply:
I think this error message would also be confusing for users. It leads to the question "what is a primary attribute?" Rather than, "oh, I guess I can't use this because it's only for functions."

Reply:
Thats a good idea, ill make the change

Reply:
I think this makes sense if we say that all primary attributes are functional attributes. The rationale for a primary attribute was to signal an attribute that cannot be applied to another item if other primary attributes were applied to it.

Perhaps its the case that currently things look like:

```
enum PrimaryAttributes{
 Functional(FunctionalAttributes)
}

enum FunctionalAttributes{
 Oracle,
 ....
}
```

Reply:
I think "PrimaryAttribute" is too general since as we expand to having more attributes, I imagine we'll have several groups of incompatible attributes rather than just the group Oracle is included in. E.g. consider a separate "pub/priv"-like group of attributes. These would be incompatible with each other but perfectly fine to use with the Oracle group. Likewise, eventually users may be able to define their own attributes with their own ad-hoc rules on where they are valid.